src/modules/job-exec/job-exec.c

```diff
@@ -399,6 +437,21 @@ static void kill_shell_timer_cb (flux_reactor_t  *r,
         job->kill_timeout = 300.;
     flux_timer_watcher_reset (w, job->kill_timeout, 0.);
     flux_watcher_start (w);
+
+    if (job->kill_shell_count >= max_kill_count
+        && job->impl->active_ranks
+        && (active_ranks = (*job->impl->active_ranks) (job))) {
+        flux_log (job->h,
+                  LOG_DEBUG,
+                  "job %s exceeded max kill count. Draining active ranks",
+                  idf58 (job->id));
+        if (drain_active_ranks (job, active_ranks) == 0) {
+            /* Force remaining tasks to be complete
+             */
+            jobinfo_tasks_complete (job, active_ranks, 1);
+        }
+        idset_destroy (active_ranks);
+    }
```

Comment:
It looks like the kill timer is re-armed even when the active ranks have been forcibly drained.  Is there a reason to do that?

Reply:
It shouldn't matter because the job entry and all associated timers will be destroyed after `jobinfo_tasks_complete()` is called. However, would be less confusing to move this conditional above the re-arm of the timer? The code would then have to return early, but only if `drain_active_ranks()` succeeds.

Reply:
Maybe add a comment explaining that it's not going to do anything if `drain_active_ranks()` succeeds but might be required if it fails?

Reply:
Ok, I added this comment, I hope that is sufficient?
```c
    /* Check if we've exceeded the maximum number of kill attempts.
     * Drain ranks that are still active if so. If the drain succeeds,
     * then force remaining tasks to be complete, which will then cause
     * this job object to be destroyed (which terminates any active kill
     * timers).
     *
     * If the drain fails (unlikely), then the job stays active and we'll
     * try to kill it again (and drain ranks) the next time the kill timer
     * fires.
     */
```

Reply:
Great!  Thanks.